### DISCUSSION TITLE
Announcements and New Features

### DISCUSSION THREAD

**Discussion by epezent** [epezent/implot#370] [Announcements]
Announcements of changes and new features will be posted here. Click `Subscribe` to the right to receive updates!

**Feel free to comment on announcements, but please do not clutter the main thread with comments.**

Previous Announcements:
[Announcements and New Features (2021)](https://github.com/epezent/implot/issues/168)
[Announcements and New Features (2020)](https://github.com/epezent/implot/issues/48)

**Comment by epezent** [comment]
## Plot Item Flags and More - (v0.14 - 6/18/2022)

### Plot Item Flags

Each of ImPlot's `PlotX` functions now takes an optional `ImPlotXFlags` parameter. This decision (1) allows us to easily add more options for plot items  and (2) allows us to consolidate and clean up some of the API. Here's what this looks like for `PlotLine`:

```cpp
// Flags for ANY PlotX function
enum ImPlotItemFlags_ {
    ImPlotItemFlags_None     = 0,
    ImPlotItemFlags_NoLegend = 1 << 0, // the item won't have a legend entry displayed
    ImPlotItemFlags_NoFit    = 1 << 1, // the item won't be considered for plot fits
};

// Flags for PlotLine
enum ImPlotLineFlags_ {
    ImPlotLineFlags_None        = 0,       // default
    ImPlotLineFlags_Segments    = 1 << 10, // a line segment will be rendered from every two consecutive points
    ImPlotLineFlags_Loop        = 1 << 11, // the last and first point will be connected to form a closed loop
    ImPlotLineFlags_SkipNaN     = 1 << 12, // NaNs values will be skipped instead of rendered as missing data
    ImPlotLineFlags_NoClip      = 1 << 13, // markers (if displayed) on the edge of a plot will not be clipped
};

...

void PlotLine(const char* label_id, const T* xs, const T* ys, int count, ImPlotLineFlags flags=0, int offset=0, int stride=sizeof(T));
```

As you can see, `PlotLine` has some new bells and whistles that are enabled with the `ImPlotLineFlags_` parameter, like rendering lines as segments or as a closed loop. Several other plotters have also gained new abilities through their respective flags (see the full list below). The special flags in `ImPlotItemFlags` can be combined with ANY `ImPlotXFlags`, e.g. `ImPlotItemFlags_NoLegend|ImPlotLineFlags_Segments` is valid code.

Here's a complete list of flags available for our current plotters:

```cpp
ImPlotItemFlags_NoLegend         // the item won't have a legend entry displayed
ImPlotItemFlags_NoFit            // the item won't be considered for plot fits

ImPlotLineFlags_Segments         // a line segment will be rendered from every two consecutive points
ImPlotLineFlags_Loop             // the last and first point will be connected to form a closed loop
ImPlotLineFlags_SkipNaN          // NaNs values will be skipped instead of rendered as missing data
ImPlotLineFlags_NoClip           // markers (if displayed) on the edge of a plot will not be clipped

ImPlotScatterFlags_NoClip        // markers on the edge of a plot will not be clipped

ImPlotStairsFlags_PreStep        // the y value is continued constantly to the left from every x position, i.e. the interval (x[i-1], x[i]] has the value y[i]

ImPlotBarsFlags_Horizontal       // bars will be rendered horizontally on the current y-axis

ImPlotBarGroupsFlags_Horizontal  // bar groups will be rendered horizontally on the current y-axis
ImPlotBarGroupsFlags_Stacked     // items in a group will be stacked on top of each other

ImPlotErrorBarsFlags_Horizontal  // error bars will be rendered horizontally on the current y-axis

ImPlotStemsFlags_Horizontal      // stems will be rendered horizontally on the current y-axis

ImPlotInfLinesFlags_Horizontal   // lines will be rendered horizontally on the current y-axis

ImPlotPieChartFlags_Normalize    // force normalization of pie chart values (i.e. always make a full circle if sum < 0)

ImPlotHeatmapFlags_ColMajor      // data will be read in column major order

ImPlotHistogramFlags_Horizontal  // histogram bars will be rendered horizontally (not supported by PlotHistogram2D)
ImPlotHistogramFlags_Cumulative  // each bin will contain its count plus the counts of all previous bins (not supported by PlotHistogram2D)
ImPlotHistogramFlags_Density     // counts will be normalized, i.e. the PDF will be visualized, or the CDF will be visualized if Cumulative is also set
ImPlotHistogramFlags_NoOutliers  // exclude values outside the specifed histogram range from the count toward normalizing and cumulative counts
ImPlotHistogramFlags_ColMajor    // data will be read in column major order (not supported by PlotHistogram)

ImPlotTextFlags_Vertical         // text will be rendered vertically
```

 **SUPER IMPORTANT NOTE**: Each `PlotX` function's `ImPlotLineFlags flags=0` parameter has been positioned before `int offset=0, int stride=sizeof(T)` where applicable. This unfortunately means that folks making use of offset and/or stride will need to update their function calls to include a flags parameter, else they will likely see strange results or crashes without any warning. This may rub some users the wrong way, but it was a necessary evil as it stylistically makes the most sense for the flags parameter to take priority over stride and offset in the default argument order for most use cases and users. 

Besides that, items flags brings about a few other **API breaking changes** users should be aware of:

- `PlotBarsH` has been removed; use `PlotBars` + `ImPlotBarsFlags_Horizontal` instead
- `PlotErrorBarsH` has been removed; use `PlotErrorBars` + `ImPlotErrorBarsFlags_Horizontal`
- `PlotVLines` and `PlotHLines` replaced with `PlotInfLines` (+ `ImPlotInfLinesFlags_Horizontal` )
- `PlotHistogram`/`PlotHistogram2D` signatures changed; `cumulative`, `density`, and `outliers` options now specified via `ImPlotHistogramFlags`
- `PlotPieChart` signature changed; `normalize` option now specified via `ImPlotPieChartFlags`
- `PlotText` signature changes; `vertical` option now specified via `ImPlotTextFlags_Vertical`

Again, this change will likely cause headaches for some users, but with this new system in place you can expect more plotting options to be added in future updates in a flexible and, importantly, maintainable way. Thanks for understanding. 

### SetupAxisScale

Previously, users could define time and log scales with `ImPlotAxisFlags_Log`, and `ImPlotAxisFlags_Time`. These flags have been replaced with a more general system using a new the setup function `SetupAxisScale`:

```cpp

// Axis scale
enum ImPlotScale_ {
    ImPlotScale_Linear = 0, // default linear scale
    ImPlotScale_Time,       // date/time scale
    ImPlotScale_Log10,      // base 10 logartithmic scale
    ImPlotScale_SymLog,     // symmetric log scale
};

// Sets an axis' scale using built-in options.
void SetupAxisScale(ImAxis axis, ImPlotScale scale);
```

`ImPlotScale_SymLog` is a new scale that produces a symmertic log scale with a linear region roughly between -1 and 1:

![image](https://user-images.githubusercontent.com/29577475/174466208-f7caf2f6-fa7e-448a-aa46-56a7280f0d3b.png)


More built-in scales may be added in the future. Until then, you can also now define **YOUR OWN** scales using `ImPlotTransform` to provide a forward and inverse function:

```cpp
// Callback signature for axis transform.
typedef double (*ImPlotTransform)(double value, void* user_data);

// Sets an axis' scale using user supplied forward and inverse transfroms.
void SetupAxisScale(ImAxis axis, ImPlotTransform forward, ImPlotTransform inverse, void* data=NULL);
```

Example:

```cpp
static inline double TransformForward_Sqrt(double v, void*) {
    return sqrt(v);
}

static inline double TransformInverse_Sqrt(double v, void*) {
    return v*v;
}

...

if (ImPlot::BeginPlot("Sqrt")) {
    ImPlot::SetupAxis(ImAxis_X1, "Linear");
    ImPlot::SetupAxis(ImAxis_Y1, "Sqrt");
    ImPlot::SetupAxisScale(ImAxis_Y1, TransformForward_Sqrt, TransformInverse_Sqrt);
    ImPlot::PlotLine(...);
    ImPlot::EndPlot();
}
```

![image](https://user-images.githubusercontent.com/29577475/174464981-e4e1afbd-b81e-4ad3-82a9-ffb3a7f8e3dd.png)

### New Axis Constraints

You can now constrain axes limits so that users can't pan beyond a min or max value, or zoom beyond a min or max width/height. This is nice if you don't want users straying away from displayed data or zooming too far into view. It is demonstrated in the candlestick demo in `implot_demo.cpp` (see Custom Plotters and Tooltips).

```cpp
// Sets an axis' limits constraints.
void SetupAxisLimitsConstraints(ImAxis axis, double v_min, double v_max);
// Sets an axis' zoom constraints.
void SetupAxisZoomConstraints(ImAxis axis, double z_min, double z_max);
```

### Misc. Improvements

- Line plots now honor ImGui's `AntiAliasedLines` and `AntiAliasedLinesUseTex`. That's right, ImPlot now uses texture based AA! If you weren't using MSAA, you can now expect signficantly smoother lines for very little performance cost.
- Markers and bar plots now render through ImPlot's custom render pipeline instead of ImGui's drawing API. This should speed up those plotters a good bit. 
- Compile times for `implot_items.cpp` should now be much quicker now.

**Comment by epezent** [comment]
# Legend Scrolling - (v0.17 - 8/20/2022)

Long legend lists will now be clipped appropriately and can be scrolled with mouse wheel. This functions for both vertical and horizontal legends, inside or outside, and legends belonging to plots or subplots. This does not currently add a physical scroll bar to long legends. I will add this feature later if users absolutely need it.

![scroll_v](https://github.com/epezent/implot/assets/29577475/f44918f5-8365-4f9a-9e12-e9373c507209)
![scroll_h](https://github.com/epezent/implot/assets/29577475/f981b73c-3147-449d-b6c4-a951833cda74)
![scroll_sub](https://github.com/epezent/implot/assets/29577475/7e1ae9f9-eefd-472c-b7b2-9433946c9bf1)

**Comment by brenocq** [comment]
# Plot Bubbles - (v1.0 - 2026-02-13)

It is now possible to plot bubbles thanks to @Luc16 in #611

```cpp
// Plots a bubble graph. #szs are the radius of each bubble in plot units.
IMPLOT_TMP void PlotBubbles(const char* label_id, const T* values, const T* szs, int count, double xscale=1, double xstart=0, ImPlotBubblesFlags flags=0, int offset=0, int stride=sizeof(T));
IMPLOT_TMP void PlotBubbles(const char* label_id, const T* xs, const T* ys, const T* szs, int count, ImPlotBubblesFlags flags=0, int offset=0, int stride=sizeof(T));
```

https://github.com/user-attachments/assets/9e6151a4-7cd4-4717-9803-dda3024ae2d6

**Comment by brenocq** [comment]
# ImPlotSpec - (v1.0 - 2026-02-14)

**⚠️ `SetNextXXX` API was deprecated in favor of `ImPlotSpec` API ⚠️**

**Usage Option 1 -- By declaring and defining a struct instance:**
```cpp
ImPlotSpec spec;
spec.LineColor = ImVec4(1,1,0,1);
spec.LineWeight = 1.0f;
spec.FillColor = ImVec4(1,0.5f,0,1);
spec.FillAlpha = 0.5f;
spec.Marker = ImPlotMarker_Square;
spec.MarkerSize = 6;
spec.Stride = sizeof(ImVec2);
spec.Flags = ImPlotItemFlags_NoLegend | ImPlotLineFlags_Shaded;
ImPlot::PlotLine("Line 1", &data1[0].x, &data1[0].y, 20, spec);
```
**Usage Option 2 -- Inline using ImPlotProp,value pairs (order does NOT matter):**
```cpp
ImPlot::PlotLine("Line 2", &data2[0].x, &data2[0].y, 20, {
    ImPlotProp_LineColor, ImVec4(0,1,1,1),
    ImPlotProp_LineWeight, 1.0f,
    ImPlotProp_FillColor, ImVec4(0,0,1,1),
    ImPlotProp_FillAlpha, 0.5f,
    ImPlotProp_Marker, ImPlotMarker_Diamond,
    ImPlotProp_MarkerSize, 6,
    ImPlotProp_Stride, sizeof(ImVec2),
    ImPlotProp_Flags, ImPlotItemFlags_NoLegend | ImPlotLineFlags_Shaded
});
```
**Result:**
![image](https://github.com/epezent/implot/assets/29577475/92e90a5e-27b7-4b4d-876f-7e0cd78de963)

**ImPlotSpec enum and structs:**
```cpp
enum ImPlotProp_ {
    ImPlotProp_LineColor,       // line color (applies to lines, bar edges); IMPLOT_AUTO_COL will use next Colormap color or current item color
    ImPlotProp_LineWeight,      // line weight in pixels (applies to lines, bar edges, marker edges)
    ImPlotProp_FillColor,       // fill color (applies to shaded regions, bar faces); IMPLOT_AUTO_COL will use next Colormap color or current item color
    ImPlotProp_FillAlpha,       // alpha multiplier (applies to FillColor and MarkerFillColor)
    ImPlotProp_Marker,          // marker type; specify ImPlotMarker_Auto to use the next unused marker
    ImPlotProp_MarkerSize,      // size of markers (radius) *in pixels*
    ImPlotProp_MarkerLineColor, // marker edge color; IMPLOT_AUTO_COL will use LineColor
    ImPlotProp_MarkerFillColor, // marker face color; IMPLOT_AUTO_COL will use LineColor
    ImPlotProp_Size,            // size of error bar whiskers (width or height), and digital bars (height) *in pixels*
    ImPlotProp_Offset,          // data index offset
    ImPlotProp_Stride,          // data stride in bytes; IMPLOT_AUTO will result in sizeof(T) where T is the type passed to PlotX
    ImPlotProp_Flags            // optional item flags; can be composed from common ImPlotItemFlags and/or specialized ImPlotXFlags
};

...

struct ImPlotSpec {
    ImVec4          LineColor       = IMPLOT_AUTO_COL;       // line color (applies to lines, bar edges); IMPLOT_AUTO_COL will use next Colormap color or current item color
    float           LineWeight      = 1.0f;                  // line weight in pixels (applies to lines, bar edges, marker edges)
    ImVec4          FillColor       = IMPLOT_AUTO_COL;       // fill color (applies to shaded regions, bar faces); IMPLOT_AUTO_COL will use next Colormap color or current item color
    float           FillAlpha       = 1.0f;                  // alpha multiplier (applies to FillColor and MarkerFillColor)
    ImPlotMarker    Marker          = ImPlotMarker_None;     // marker type; specify ImPlotMarker_Auto to use the next unused marker
    float           MarkerSize      = 4;                     // size of markers (radius) *in pixels*
    ImVec4          MarkerLineColor = IMPLOT_AUTO_COL;       // marker edge color; IMPLOT_AUTO_COL will use LineColor
    ImVec4          MarkerFillColor = IMPLOT_AUTO_COL;       // marker face color; IMPLOT_AUTO_COL will use LineColor
    float           Size            = 4;                     // size of error bar whiskers (width or height), and digital bars (height) *in pixels*
    int             Offset          = 0;                     // data index offset
    int             Stride          = IMPLOT_AUTO;           // data stride in bytes; IMPLOT_AUTO will result in sizeof(T) where T is the type passed to PlotX
    ImPlotItemFlags Flags           = ImPlotItemFlags_None;  // optional item flags; can be composed from common ImPlotItemFlags and/or specialized ImPlotXFlags

     ...
}
```

## `SetNextXXX` Migration Guide
- `SetNextLineStyle` has been removed, styling should be set via ImPlotSpec.
  ```cpp
  // Before
  ImPlot::SetNextLineStyle(line_color, line_weight);
  ImPlot::PlotLine("Line", xs, ys, count);
  
  // After
  ImPlotSpec spec;
  spec.LineColor = line_color;
  spec.LineWeight = line_weight;
  ImPlot::PlotLine("Line", xs, ys, count, spec);
  ```
- `SetNextFillStyle` has been removed, styling should be set via ImPlotSpec.
  ```cpp
  // Before
  ImPlot::SetNextFillStyle(fill_color, fill_alpha);
  ImPlot::PlotLine("Shaded", xs, ys, count, ImPlotLineFlags_Shaded);

  // After
  ImPlotSpec spec;
  spec.FillColor = fill_color;
  spec.FillAlpha = fill_alpha;
  spec.Flags = ImPlotLineFlags_Shaded;
  ImPlot::PlotTLine("Shaded", xs, ys, count, spec);
  ```
- `SetNextMarkerStyle` has been removed, styling should be set via ImPlotSpec.
   ```cpp
   // Before
   ImPlot::SetNextMarkerStyle(marker, marker_size, fill_color, line_weight, marker_outline_color);
   ImPlot::PlotScatter("Scatter", xs, ys, count);

   // After
   ImPlotSpec spec;
   spec.LineWeight = line_weight;
   spec.Marker = marker;
   spec.MarkerSize = marker_size;
   spec.MarkerLineColor = marker_outline_color;
   spec.MarkerFillColor = fill_color;
   ImPlot::PlotScatter("Scatter", xs, ys, count, spec);
   ```
- `SetNextErrorBarStyle` has been removed, styling should be set via ImPlotSpec.
   ```cpp
   // Before
   ImPlot::SetNextErrorBarStyle(color, size, weight);
   ImPlot::PlotErrorBars("ErrorBar", xs, ys, err, count);

   // After
   ImPlotSpec spec;
   spec.LineColor = color;
   spec.Size = size;
   spec.LineWeight = weight;
   ImPlot::PlotErrorBars("ErrorBar", xs, ys, err, count, spec);
   ```

- Flags, Offset and Stride should also be set via ImPlotSpec now.

**Comment by brenocq** [comment]
# Plot Polygon -  (v1.0 - 2026-03-07)

It is now possible to plot polygons! Requested by @ssteinbach in #631, implemented in #673.

```cpp
// Plots a polygon. Points are specified in counter-clockwise order. If concave, make sure to set the Concave flag.
IMPLOT_TMP void PlotPolygon(const char* label_id, const T* xs, const T* ys, int count, const ImPlotSpec& spec=ImPlotSpec());
```

```cpp
if (ImPlot::BeginPlot("Polygon Plot", ImVec2(-1,0), ImPlotFlags_Equal)) {
    ImPlot::PlotPolygon("Triangle", tri_xs, tri_ys, 3, {
        ImPlotProp_FillAlpha, 0.5f,
    });
    ImPlot::PlotPolygon("Pentagon", pent_xs, pent_ys, 5, {
        ImPlotProp_FillAlpha, 0.5f,
        ImPlotProp_FillColor, ImVec4(0,1,0,1),
    });
    ImPlot::PlotPolygon("Star (Concave)", star_xs, star_ys, 10, {
        ImPlotProp_FillAlpha, 0.5f,
        ImPlotProp_FillColor, ImVec4(1,1,0,1),
        ImPlotProp_Flags, ImPlotPolygonFlags_Concave,
    });

    ImPlot::EndPlot();
}
```
<img width="606" height="325" alt="image" src="https://github.com/user-attachments/assets/eb28ec60-93b8-4bd1-ae6e-04eb58d83f8a" />

**Comment by brenocq** [comment]
# Per-Index Color/Size - (v1.0 - 2026-04-04)

<p align="center">
<img width="300" alt="image" src="https://github.com/user-attachments/assets/85414f8d-03dc-4064-8b9c-325eccc4a16d" />
<img width="300" alt="image" src="https://github.com/user-attachments/assets/fdd61d1e-9621-4c1f-b9e8-5bd2d7cd2d45" />
</p>

<p align="center">
<img width="300" alt="image" src="https://github.com/user-attachments/assets/62333f2d-9938-44c9-91a5-af746660ebf2" />
<img width="300" alt="image" src="https://github.com/user-attachments/assets/413d9ffd-a191-4156-9cd6-f60e37d6aea1" />
</p>


Five new `ImPlotSpec` properties have been added to support per-index coloring and sizing across most plot types:

```cpp
ImPlotProp_LineColors,        // ImU32* array of colors for each line segment/vertex
ImPlotProp_FillColors,        // ImU32* array of colors for each filled region
ImPlotProp_MarkerLineColors,  // ImU32* array of colors for each marker outline
ImPlotProp_MarkerFillColors,  // ImU32* array of colors for each marker face
ImPlotProp_MarkerSizes,       // float* array of sizes for each marker
```

When not specified, plots fall back to their existing single-color properties — no breaking changes.

**Example — rainbow line:**
```cpp
static ImU32 colors[100];
for (int i = 0; i < 100; ++i)
    colors[i] = ImColor::HSV(i / 99.0f, 0.8f, 0.9f);

ImPlot::PlotLine("f(x)", xs, ys, 100, {
    ImPlotProp_LineColors, colors
});
```

**Example — scatter with per-marker color and size:**
```cpp
ImPlot::PlotScatter("Data", xs, ys, 50, {
    ImPlotProp_MarkerFillColors, fill_colors,
    ImPlotProp_MarkerLineColors, line_colors,
    ImPlotProp_MarkerSizes,      sizes
});
```

**Example — colored bars:**
```cpp
ImPlot::PlotBars("Bars", data, 10, 0.7, 1, {
    ImPlotProp_FillColors, colors,
    ImPlotProp_LineColors, colors
});
```

Supported plot types: `PlotLine`, `PlotScatter`, `PlotBars`, `PlotShaded`, `PlotBubbles`, `PlotStairs`, `PlotStems`, `PlotInfLines`.

Colors are specified as `ImU32` — use `ImGui::GetColorU32()` or `ImColor` to convert from `ImVec4`. `FillAlpha` is applied on top of per-index alpha values when set.
